moby/moby
# set of fixes for btrfs graphdriver

**- What I did**
Fixes #35630 btrfs graphdriver keeps subvolumes in readonly state while they not used by `Get`.
Rework subvolumes delete function in more efficient way.
Fix quotas usage to handle cases while file system shared with others.

**- How I did it**
I add subvolSetPropRO function to btrfs graphdriver which could set readonly property to subvolume, and call it at Get or Put.
Rework subvolDelete to increase performance by using BTRFS_IOC_TREE_SEARCH and left a full walk of the directory tree as a last resort. Also handle paths longer than PATH_MAX.
Eliminate in-process quota state. Remove ability to disable quotas because only user know when it's save. Invoke quotas rescan when it's really necessary.

**- How to verify it**
Intermediate layers in /var/lib/docker/btrfs/subvolumes/ are read-only. Try to write something and you'll see.
Enable quotas before daemon start or disable while it's running. Qgroups not left after deleted subvols, quota rescan normally not happen.

**- Description for the changelog**
btrfs graphdriver keeps subvolumes in read-only state when possible
